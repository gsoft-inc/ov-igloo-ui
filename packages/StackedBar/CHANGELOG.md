# @igloo-ui/stacked-bar

## 0.2.1

### Patch Changes

- 9739155: Updated storybook and recharts. Applied the new brand look for filter.
- 9739155: Gave ColorPicker the new Wokleap look. Color now uses css variables to update its color. Fixed spacing issues in tooltip. Updated hopper token version.
- Updated dependencies [9739155]
  - @igloo-ui/popover@0.4.1

## 0.2.0

### Minor Changes

- 4096378: Added the rebrand. Added disabled prop to Popover to work with StackedBar.

### Patch Changes

- Updated dependencies [4096378]
  - @igloo-ui/popover@0.4.0

## 0.1.4

### Patch Changes

- bf4ddef: Updated storybook version and linting. Made linting fixes in these components.
- Updated dependencies [bf4ddef]
  - @igloo-ui/tooltip@3.1.6

## 0.1.3

### Patch Changes

- 89d10c5: Updated recharts version and fixed dropdown mock

## 0.1.2

### Patch Changes

- a41e1d4: Fix: moved destructuring of props to function parameters so that storybook will automatically show the default value in the props table. It was done in Modal because it needed to send props to useDialog. In this case, @default annotation was used in comments.
- Updated dependencies [a41e1d4]
  - @igloo-ui/tooltip@3.1.3

## 0.1.1

### Patch Changes

- 00fdb02: Updated tooltip class and hid the arrow on the stacked bar tooltip.
- bf478b5: Added missing tests. Fixed missing code in stories when clicking 'Show code'. Updated Dropdown mock. Added new optional prop highlightToday to help with test snapshots.
- Updated dependencies [bf478b5]
  - @igloo-ui/tooltip@3.1.2

## 0.1.0

### Minor Changes

- 49d3c6b: The initial release of the stacked bar component

### Patch Changes

- Updated dependencies [49d3c6b]
  - @igloo-ui/tooltip@3.1.1
