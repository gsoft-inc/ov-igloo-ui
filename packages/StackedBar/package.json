{
  "name": "@igloo-ui/stacked-bar",
  "version": "0.2.0",
  "main": "dist/StackedBar.js",
  "module": "dist/StackedBar.js",
  "types": "dist/StackedBar.d.ts",
  "styles": "dist/stacked-bar.css",
  "publishConfig": {
    "access": "public"
  },
  "repository": "git+https://github.com/gsoft-inc/ov-igloo-ui.git",
  "author": "Officevibe Inc.",
  "contributors": [
    "Vicky Comeau vicky.comeau@gsoft.com"
  ],
  "license": "Apache-2.0",
  "peerDependencies": {
    "react": ">= 16.8.6"
  },
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "rollup -c rollup.config.js"
  },
  "dependencies": {
    "@igloo-ui/popover": "^0.4.0",
    "@igloo-ui/tokens": "^2.0.0",
    "@hopper-ui/tokens": "^0.3.0",
    "classnames": "^2.3.2",
    "recharts": "^2.9.2"
  },
  "browserslist": [
    "> 1%",
    "last 2 versions",
    "IE >= 11"
  ]
}
