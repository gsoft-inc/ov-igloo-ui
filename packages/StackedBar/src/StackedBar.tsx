import * as React from 'react';
import cx from 'classnames';
import {
  BarChart,
  XAxis,
  YAxis,
  Bar,
  XAxisProps,
  YAxisProps,
  ResponsiveContainer,
  LabelProps,
} from 'recharts';
// eslint-disable-next-line max-len
import { CategoricalChartProps } from 'recharts/types/chart/generateCategoricalChart';
import variables from '@igloo-ui/tokens/dist/base10/tokens.json';

import Tooltip from '@igloo-ui/tooltip';

import './stacked-bar.scss';
import StackedBarTooltip from './StackedBarTooltip';

export interface StackedBarLabel extends LabelProps {
  dataKey?: string;
}

export type Pos = 'first' | 'last';

export type Strength = -2 | -1 | 0 | 1 | 2;

interface ValueRange {
  /** The min value of the x axis */
  min: number | 'auto' | 'dataMin';
  /** The min value of the x axis */
  max: number | 'auto' | 'dataMax';
}

export interface DataSet {
  /** Set a custom Hex or RGB color */
  color?: string;
  /** The key that represents the stack on the bar */
  key: string;
  /** The label that will be displayed on the tooltip */
  label: string;
  /** Set a number representing the value of the stack */
  value: number;
  /** This indicates the strength on the scale and
   * it will automatically set the color if not overridden */
  strength?: Strength;
}

export interface StackedBarProps extends React.ComponentProps<'div'> {
  /** Add a class name to the stacked bar */
  className?: string;
  /** All the data needed to build the stacked bar. If empty, leave blank */
  dataSet?: DataSet[];
  /** Add a data-test tag for automated tests */
  dataTest?: string;
  /** A function to format the value in the tooltip (Default: adds %) */
  formatValue?: (value: number) => string;
  /** Message to display when there is no data */
  noDataMessage?: string;
  /** Show the values on the bar */
  showValue?: boolean;
  /** Adjusts the height of the bar */
  size?: 'small' | 'medium';
  /** The range that the domain of the value should be based on */
  valueRange?: ValueRange;
}

const StackedBar: React.FunctionComponent<StackedBarProps> = ({
  className,
  dataSet,
  dataTest,
  formatValue = (value: number) => {
    return `${value}%`;
  },
  noDataMessage,
  showValue = false,
  size = 'medium',
  valueRange = { min: 0, max: 100 },
  ...rest
}: StackedBarProps) => {
  const xAxisConfig: XAxisProps = {
    type: 'number',
    hide: true,
    domain: [valueRange.min, valueRange.max],
  };

  const yAxisConfig: YAxisProps = {
    type: 'category',
    hide: true,
  };

  const barChartConfig: CategoricalChartProps = {
    layout: 'vertical',
    barCategoryGap: '0%',
    margin: { top: 0, right: 0, bottom: 0, left: 0 },
  };

  const barConfig = {
    stackId: 'a',
    isAnimationActive: false,
  };

  const getLabel = (dataKey: string): StackedBarLabel => ({
    dataKey,
    formatter: (value: number) =>
      showValue && value && size === 'medium' ? value : null,
    className: 'ids-stacked-bar__label',
  });

  const getRadius = (position?: Pos): [number, number, number, number] => {
    return [
      position === 'first' ? 4 : 0,
      position === 'last' ? 4 : 0,
      position === 'last' ? 4 : 0,
      position === 'first' ? 4 : 0,
    ];
  };

  const getColor = (dataItem: DataSet): string => {
    if (dataItem.color) {
      return dataItem.color;
    }

    if (dataItem.strength !== undefined) {
      switch (dataItem.strength) {
        case -2:
          return variables.coral400;
        case -1:
          return variables.coral200;
        case 1:
          return variables.seaweed200;
        case 2:
          return variables.seaweed400;
        default:
          return variables.creamsicle100;
      }
    }

    return variables.grey300;
  };

  const emptyBars = [
    <Bar
      key="empty-bar-1"
      {...barConfig}
      dataKey="firstBar"
      className="ids-stacked-bar__empty-bar--side"
      radius={[4, 0, 0, 4]}
      filter="url(#inset-shadow)"
    />,
    <Bar
      key="empty-bar-2"
      {...barConfig}
      dataKey="secondBar"
      className="ids-stacked-bar__empty-bar--middle"
      filter="url(#inset-shadow)"
    />,
    <Bar
      key="empty-bar-3"
      {...barConfig}
      dataKey="thirdBar"
      className="ids-stacked-bar__empty-bar--side"
      radius={[0, 4, 4, 0]}
      filter="url(#inset-shadow)"
    />,
  ];

  const barShadow = (
    <defs>
      <filter id="inset-shadow">
        <feOffset dx="0" dy="2" />
        <feGaussianBlur stdDeviation="2" result="offset-blur" />
        <feComposite
          operator="out"
          in="SourceGraphic"
          in2="offset-blur"
          result="inverse"
        />
        <feFlood floodColor="#253346" floodOpacity=".10" result="color" />
        <feComposite operator="in" in="color" in2="inverse" result="shadow" />
        <feComposite operator="over" in="shadow" in2="SourceGraphic" />
      </filter>
    </defs>
  );

  const barDataObj: Record<string, unknown> = {};
  let bars = null;

  if (dataSet && dataSet.length) {
    bars = dataSet.map((barInfo, key) => {
      if (barInfo.value) {
        const dataKey = barInfo.key;
        let pos;

        if (key === 0) {
          pos = 'first' as Pos;
        }

        if (key === dataSet.length - 1) {
          pos = 'last' as Pos;
        }

        barDataObj[dataKey] = barInfo.value;

        return (
          <Bar
            {...barConfig}
            key={`data-bar-${dataKey}`}
            className={cx(
              'ids-stacked-bar__bar',
              `ids-stacked-bar__bar--${dataKey}`,
              {
                'ids-stacked-bar__bar--is-value-visible': showValue,
                [`ids-stacked-bar__bar--strength-${barInfo.strength}`]:
                  barInfo.strength !== undefined,
                'ids-stacked-bar__bar--no-color':
                  barInfo.strength === undefined && !barInfo.color,
              },
            )}
            dataKey={dataKey}
            label={getLabel(dataKey)}
            fill={getColor(barInfo)}
            radius={getRadius(pos)}
          />
        );
      }

      return null;
    });
  }

  const hasData = !!bars;

  const barChart = (() => {
    if (hasData) {
      return {
        data: [barDataObj],
        component: bars,
        tooltip: (
          <StackedBarTooltip
            hasData={hasData}
            dataSet={dataSet}
            formatValue={formatValue}
          />
        ),
      };
    }

    return {
      data: [
        {
          firstBar: 33,
          secondBar: 34,
          thirdBar: 33,
        },
      ],
      component: emptyBars,
      tooltip: (
        <StackedBarTooltip noDataMessage={noDataMessage} hasData={hasData} />
      ),
    };
  })();

  const classes = cx('ids-stacked-bar', className, `ids-stacked-bar--${size}`);

  return (
    <div className={classes} data-test={dataTest} {...rest}>
      <Tooltip
        content={barChart.tooltip}
        appearance="light"
        className="ids-stacked-bar-tooltip__container"
        tooltipClassName="ids-stacked-bar-tooltip"
        disabled={!hasData && !noDataMessage}
        maxWidth={600}
      >
        <ResponsiveContainer width="100%" height="100%">
          <BarChart
            data={barChart.data}
            className="ids-stacked-bar__chart"
            {...barChartConfig}
          >
            {!hasData && barShadow}
            <XAxis {...xAxisConfig} />
            <YAxis {...yAxisConfig} />
            {barChart.component}
          </BarChart>
        </ResponsiveContainer>
      </Tooltip>
    </div>
  );
};

export default StackedBar;
