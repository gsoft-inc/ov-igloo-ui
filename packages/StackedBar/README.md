# StackedBar

A stacked bar is a representation of data that uses rectangular bars to show the relative proportions of different categories.

<Example is="custom" />

<ReferenceLinks is="custom" />

## Installation

To install `@igloo-ui/stacked-bar` in your project, you will need to run the following command using [npm](https://www.npmjs.com/):

```bash
npm install @igloo-ui/stacked-bar
```

If you prefer [Yarn](https://classic.yarnpkg.com/en/), use the following command instead:

```bash
yarn add @igloo-ui/stacked-bar
```

## Usage

Then to use the component in your code just import it!

```jsx
import StackedBar from '@igloo-ui/stacked-bar';

<StackedBar
  dataSet={[
    {
      key: 'stronglyUnfavorable',
      label: 'Strongly Unfavorable',
      value: 30,
      strength: -2,
    },
    {
      key: 'unfavorable',
      label: 'Unfavorable',
      value: 10,
      strength: -1,
    },
    {
      key: 'neutral',
      label: 'Neutral',
      value: 20,
      strength: 0,
    },
    {
      key: 'favorable',
      label: 'Favorable',
      value: 15,
      strength: 1,
    },
    {
      key: 'stronglyFavorable',
      label: 'Strongly Favorable',
      value: 25,
      strength: 2,
    },
  ]}
  showValue
/>;
```
