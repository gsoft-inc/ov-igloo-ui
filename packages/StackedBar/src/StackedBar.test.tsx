/**
 * @jest-environment jsdom
 */
import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { ResponsiveContainerProps } from 'recharts';
import MockTooltip from '@igloo-ui/tooltip/src/__mocks__/Tooltip.mock';

import StackedBar from './StackedBar';

const label = 'Strongly Unfavorable';
const dataSet = [
  {
    key: 'stronglyUnfavorable',
    label: label,
    value: 30,
    strength: -2,
  },
];

const setup = (props = {}) => {
  return render(<StackedBar dataTest="ids-stacked-bar" {...props} />);
};

jest.mock('@igloo-ui/tooltip', () => ({
  __esModule: true,
  default: jest.fn(MockTooltip),
}));

jest.mock('recharts', () => {
  const OriginalModule = jest.requireActual('recharts');

  return {
    ...OriginalModule,
    ResponsiveContainer: ({ children }: ResponsiveContainerProps) => (
      <OriginalModule.ResponsiveContainer width={600} aspect={1}>
        {children}
      </OriginalModule.ResponsiveContainer>
    ),
  };
});

describe('StackedBar', () => {
  test('It should render without errors', () => {
    setup();
    const wrapper = screen.getByTestId('ids-stacked-bar');
    const chart = screen.getByRole('region');
    expect(wrapper).toBeInTheDocument();
    expect(chart).toBeInTheDocument();
  });

  test('It should render a snapshot', () => {
    const { asFragment } = setup();
    expect(asFragment()).toMatchSnapshot();
  });

  test('It should render a medium size by default', () => {
    setup();
    const wrapper = screen.getByTestId('ids-stacked-bar');
    expect(wrapper).toHaveClass('ids-stacked-bar--medium');
  });

  test('It should render a medium size', () => {
    setup({ size: 'medium' });
    const wrapper = screen.getByTestId('ids-stacked-bar');
    expect(wrapper).toHaveClass('ids-stacked-bar--medium');
  });

  test('It should render a small size', () => {
    setup({ size: 'small' });
    const wrapper = screen.getByTestId('ids-stacked-bar');
    expect(wrapper).toHaveClass('ids-stacked-bar--small');
  });

  test('It should render a no data tooltip', () => {
    const message = 'No data';
    const { container } = setup({ noDataMessage: message });
    const tooltip = container.querySelector(
      '.ids-stacked-bar-tooltip__container'
    );
    if (tooltip) {
      fireEvent.click(tooltip);
    }
    expect(screen.getByText(message)).toBeInTheDocument();
  });

  test('It should render a tooltip with data', () => {
    const { container } = setup({ dataSet: dataSet });
    const tooltip = container.querySelector(
      '.ids-stacked-bar-tooltip__container'
    );
    if (tooltip) {
      fireEvent.click(tooltip);
    }
    expect(screen.getByText(label)).toBeInTheDocument();
  });

  test('It should render a tooltip with a custom format', () => {
    const { container } = setup({
      dataSet: dataSet,
      formatValue: (value: number) => {
        return `${value} units`;
      },
    });
    const tooltip = container.querySelector(
      '.ids-stacked-bar-tooltip__container'
    );
    if (tooltip) {
      fireEvent.click(tooltip);
    }
    expect(screen.getByText('30 units')).toBeInTheDocument();
  });

  test('It should add a class for strength', () => {
    const { container } = setup({
      dataSet: dataSet,
      formatValue: (value: number) => {
        return `${value} units`;
      },
    });
    const bar = container.querySelector('.ids-stacked-bar__bar');

    expect(bar).toHaveClass('ids-stacked-bar__bar--strength--2');
  });
});
