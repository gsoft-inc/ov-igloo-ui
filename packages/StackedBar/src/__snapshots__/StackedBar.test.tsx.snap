// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StackedBar It should render a snapshot 1`] = `
<DocumentFragment>
  <div
    class="ids-stacked-bar ids-stacked-bar--medium"
    data-test="ids-stacked-bar"
  >
    <span
      class="ids-tooltip__container ids-stacked-bar-tooltip__container"
      role="button"
      tabindex="0"
    >
      <div
        class="recharts-responsive-container"
        style="width: 600px; height: 100%;"
      >
        <div
          class="recharts-wrapper ids-stacked-bar__chart"
          role="region"
          style="position: relative; cursor: default; width: 600px; height: 600px;"
        >
          <svg
            class="recharts-surface"
            height="600"
            viewBox="0 0 600 600"
            width="600"
          >
            <title />
            <desc />
            <defs>
              <clippath
                id="recharts5-clip"
              >
                <rect
                  height="600"
                  width="600"
                  x="0"
                  y="0"
                />
              </clippath>
            </defs>
            <defs>
              <filter
                id="inset-shadow"
              >
                <feoffset
                  dx="0"
                  dy="2"
                />
                <fegaussianblur
                  result="offset-blur"
                  stdDeviation="2"
                />
                <fecomposite
                  in="SourceGraphic"
                  in2="offset-blur"
                  operator="out"
                  result="inverse"
                />
                <feflood
                  flood-color="#253346"
                  flood-opacity=".10"
                  result="color"
                />
                <fecomposite
                  in="color"
                  in2="inverse"
                  operator="in"
                  result="shadow"
                />
                <fecomposite
                  in="shadow"
                  in2="SourceGraphic"
                  operator="over"
                />
              </filter>
            </defs>
            <g
              class="recharts-layer recharts-bar ids-stacked-bar__empty-bar--side"
            >
              <g
                class="recharts-layer recharts-bar-rectangles"
              >
                <g
                  class="recharts-layer recharts-bar-rectangle"
                  role="img"
                >
                  <path
                    class="recharts-rectangle ids-stacked-bar__empty-bar--side"
                    d="M0,4A 4,4,0,0,1,4,0L 198,0L 198,600L 4,600A 4,4,0,0,1,
        0,596Z"
                    filter="url(#inset-shadow)"
                    height="600"
                    radius="4,0,0,4"
                    width="198"
                    x="0"
                    y="0"
                  />
                </g>
              </g>
              <g
                class="recharts-layer"
              />
            </g>
            <g
              class="recharts-layer recharts-bar ids-stacked-bar__empty-bar--middle"
            >
              <g
                class="recharts-layer recharts-bar-rectangles"
              >
                <g
                  class="recharts-layer recharts-bar-rectangle"
                  role="img"
                >
                  <path
                    class="recharts-rectangle ids-stacked-bar__empty-bar--middle"
                    d="M 198,0 h 204 v 600 h -204 Z"
                    filter="url(#inset-shadow)"
                    height="600"
                    radius="0"
                    width="204"
                    x="198"
                    y="0"
                  />
                </g>
              </g>
              <g
                class="recharts-layer"
              />
            </g>
            <g
              class="recharts-layer recharts-bar ids-stacked-bar__empty-bar--side"
            >
              <g
                class="recharts-layer recharts-bar-rectangles"
              >
                <g
                  class="recharts-layer recharts-bar-rectangle"
                  role="img"
                >
                  <path
                    class="recharts-rectangle ids-stacked-bar__empty-bar--side"
                    d="M402,0L 596,0A 4,4,0,0,1,
        600,4L 600,596A 4,4,0,0,1,
        596,600L 402,600Z"
                    filter="url(#inset-shadow)"
                    height="600"
                    radius="0,4,4,0"
                    width="198"
                    x="402"
                    y="0"
                  />
                </g>
              </g>
              <g
                class="recharts-layer"
              />
            </g>
          </svg>
        </div>
      </div>
    </span>
  </div>
</DocumentFragment>
`;
